### ISSUE TITLE
Train on something different than text

### ISSUE THREAD

**Issue by ethancohen123** [lucidrains/DALLE2-pytorch#232]
Hi,
Is it possible to train something different than text to image ? Im thinking of something like features or other to image. 
x-clip is already working for that as we discussed in this issue : https://github.com/lucidrains/x-clip/issues/12
But it seems like the diffusion prior is only working with text--> images. Is there a way to make it possible to condition of other stuffs ? 
Thanks :)

**Comment by Veldrovive** [comment]
The prior needs to be retrained for any new clip model being used whether or not it is encoding something other than text. If you have a clip model that is encoding something other than text it would be possible to use that, but you would need to train a prior and decoder for that task specifically and the data requirements remain very large.

**Comment by ethancohen123** [comment]
Yep I trained a new clip and decoder but the diffusion prior seems to only work with text not anything else

**Comment by ethancohen123** [comment]
Anyway to go around that easily ? @lucidrains

**Comment by Veldrovive** [comment]
Did you train a new prior?

**Comment by ethancohen123** [comment]
well I have trained clip and the decoder. Then I tried to train the prior network and I got the following error: 
mat1 and mat2 shapes cannot be multiplied (2x256 and 15724x2048) with a traceback on encoder_output = self.clip.text_transformer(text)

**Comment by Veldrovive** [comment]
I would assume that is due to the dimension of the embedding being incorrect. What is the shape of the output from your clip?

**Comment by ethancohen123** [comment]
output of clip is of size (batch_size) (=2 in my code for testing)

**Comment by Veldrovive** [comment]
Oh, I didn't read carefully enough. This is a problem in the XClipAdapter. Maybe a text length issue? Did you use this clip to train the decoder?

**Comment by ethancohen123** [comment]
here is a full test code snippet 


```
import torch
from dalle2_pytorch import CLIP

import torch
import torch.nn as nn
from torchvision import models
from einops import rearrange


class features_encoder(torch.nn.Module):
    #output size is (bs,512)
    def __init__(self):
        super(features_encoder, self).__init__()
        
        self.model=nn.Sequential(
            nn.Linear(15724,2048),
            nn.ReLU(),
            nn.Linear(2048,512)
            
        )
        
    def forward(self, x):
        out=self.model(x)
        return out


features_encoder=features_encoder().cuda()

clip = CLIP(
    text_encoder = features_encoder,
    text_encode_without_mask = True,
    dim_image = 512,
    dim_text = 512,
    dim_latent = 512
).cuda()

images=torch.randn((2,3,256,256)).cuda()
feat=torch.randn((2,15724)).cuda()

loss = clip(feat, images, return_loss =True)
loss.backward()


import torch
from dalle2_pytorch import Unet, Decoder, CLIP

unet = Unet(
    dim = 128,
    image_embed_dim = 512,
    cond_dim = 128,
    channels = 3,
    dim_mults=(1, 2, 4, 8)
).cuda()

# decoder, which contains the unet and clip

decoder = Decoder(
    unet = unet,
    clip = clip,
    timesteps = 100,
    image_cond_drop_prob = 0.1,
    text_cond_drop_prob = 0.5
).cuda()


images=torch.randn((2,3,256,256)).cuda()

loss = decoder(images)
loss.backward()

from dalle2_pytorch import DiffusionPriorNetwork, DiffusionPrior, CLIP


# setup prior network, which contains an autoregressive transformer

prior_network = DiffusionPriorNetwork(
    dim = 512,
    depth = 6,
    dim_head = 64,
    heads = 8
).cuda()

# diffusion prior network, which contains the CLIP and network (with transformer) above

diffusion_prior = DiffusionPrior(
    net = prior_network,
    clip = clip,
    timesteps = 100,
    cond_drop_prob = 0.2
).cuda()


images=torch.randn((2,3,256,256)).cuda()
feat=torch.randn((2,15724)).cuda()

loss = diffusion_prior(feat,images)
loss.backward()
```

**Comment by ethancohen123** [comment]
The bug happens in the last part aka loss=diffusion_prior

**Comment by Veldrovive** [comment]
This is actually an issue with the x-clip configuration. I can't dig deep right now, but I think there are a couple problems. The most obvious one is that you need to inform clip that it will be taking long sequences. By default it expects the text sequence to be under 256 long while yours is 15724 and the way your text encoder was built it always expects exactly 15724. You can set the max text size with the `text_seq_len`, but I think there will be further issues. @lucidrains  is the expert on x-clip and I know basically nothing so he will be much more able to help.

**Comment by lucidrains** [comment]
@Veldrovive thanks for fielding this :)

@ethancohen123 so the issue is that you are projecting your text along the sequence length to the latent directly, so there is no text encodings per se

try upgrading to 1.8.3, then use this modified script

```python
import torch
from dalle2_pytorch import CLIP

import torch
import torch.nn as nn
from torchvision import models
from einops import rearrange


class features_encoder(torch.nn.Module):
    #output size is (bs,512)
    def __init__(self):
        super(features_encoder, self).__init__()
        
        self.model=nn.Sequential(
            nn.Linear(15724,2048),
            nn.ReLU(),
            nn.Linear(2048,512)
            
        )
        
    def forward(self, x):
        out=self.model(x)
        return out


features_encoder=features_encoder().cuda()

clip = CLIP(
    text_encoder = features_encoder,
    text_encode_without_mask = True,
    text_seq_len = 15724,                       # <-- this needs to be set to the correct maximum text sequence length
    dim_image = 512,
    dim_text = 512,
    dim_latent = 512
).cuda()

images=torch.randn((2,3,256,256)).cuda()
feat=torch.randn((2,15724)).cuda()

loss = clip(feat, images, return_loss =True)
loss.backward()


import torch
from dalle2_pytorch import Unet, Decoder, CLIP

unet = Unet(
    dim = 16,
    image_embed_dim = 512,
    cond_dim = 128,
    channels = 3,
    dim_mults=(1, 2, 4, 8)
).cuda()

# decoder, which contains the unet and clip

decoder = Decoder(
    unet = unet,
    clip = clip,
    timesteps = 100,
    image_cond_drop_prob = 0.1,
    text_cond_drop_prob = 0.5
).cuda()


images=torch.randn((2,3,256,256)).cuda()

loss = decoder(images)
loss.backward()

from dalle2_pytorch import DiffusionPriorNetwork, DiffusionPrior, CLIP


# setup prior network, which contains an autoregressive transformer

prior_network = DiffusionPriorNetwork(
    dim = 512,
    depth = 6,
    dim_head = 64,
    heads = 8
).cuda()

# diffusion prior network, which contains the CLIP and network (with transformer) above

diffusion_prior = DiffusionPrior(
    net = prior_network,
    clip = clip,
    timesteps = 100,
    cond_drop_prob = 0.2,
    condition_on_text_encodings = False     # <-- set this to False
).cuda()


images=torch.randn((2,3,256,256)).cuda()
feat=torch.randn((2,15724)).cuda()

loss = diffusion_prior(feat,images)
loss.backward()
```

**Comment by lucidrains** [comment]
@ethancohen123 just curious, but what kind of modality are you working with?

**Comment by ethancohen123** [comment]
Thanks i will try this during the we !
Working with laion medical on brain human interface problem :)

**Comment by ethancohen123** [comment]
But also working on different multi modal generation but different than text to image, we could have a chat about this if youd like :)

**Comment by lucidrains** [comment]
@ethancohen123 cool! yea, maybe once i finish what is on my plate in a couple months we can chat :)
